### PR TITLE
Add multiple new API endpoints

### PR DESCRIPTION
**Pull request by mommothazaz123**
Added endpoints:
`GET /api/user?username=:un`
`POST /api/character/:_id/spellList/:listId`
`POST /api/character`
`DELETE /api/character/:_id`
`PUT /api/character/:_id/owner`
`POST /api/character/:_id/feature`
`POST /api/character/:_id/prof`
`POST /api/character/:_id/effect`
`POST /api/character/:_id/class`

All endpoints require an API key, do correct permission checking, and have corresponding ratelimits (all currently set to 5/5, feel free to change these).

Additionally, allowed the server to write to a collection no matter what, as operations originating from the server were being blocked under the pretence of "No UserId set when trying to update character asset."

This change should massively decrease the amount of mucking about with the websocket required, as well as reduce the number of individual calls; the endpoints allow for mass updating, requiring only one API call to insert `n` spells, rather than `n`.

### CODE REVIEW

File: app/Routes/API.js [hunk 1]
```diff
@@ -1,91 +1,391 @@
...
+            var key = query && query.key;
+            ifKeyValid(key, this.response, "vmixCharacter", () =>
+                this.response.end(vMixCharacter(this.params._id))
+            );
+        },
+    });
+    this.route("vmixParty", {
+        path: "/vmix-party/:_id/",
+        where: "server",
+        action: function () {
```
Comment (ThaumRystra) [inline app/Routes/API.js]: Actions should each just end up being `Meteor.call('someMethod', args)` with all the validation in there, which will be considered trusted code, and will always be allowed

Comment (mommothazaz123) [reply app/Routes/API.js]: Currently to do the validation in a Meteor method I'd have to pass either the Router or Response object to the Meteor method, which throws a RangeError (is it trying to serialize the object to EJSON?). Are there any alternates to this, or do I have to rewrite all the validation code?

Comment (mommothazaz123) [reply app/Routes/API.js]: Traceback: 
```
W20190207-16:39:23.935(-8)? (STDERR) RangeError: Maximum call stack size exceeded
W20190207-16:39:23.937(-8)? (STDERR)     at Array.forEach (<anonymous>)
W20190207-16:39:23.938(-8)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190207-16:39:23.938(-8)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
[...repeats a few times]
W20190207-16:39:23.946(-8)? (STDERR)  => awaited here:
W20190207-16:39:23.947(-8)? (STDERR)     at Promise.await (/Users/andrew/.meteor/packages/promise/.0.11.1.1calf8t.hslz++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/promise_server.js:60:12)
W20190207-16:39:23.947(-8)? (STDERR)     at Server.apply (packages/ddp-server/livedata_server.js:1634:14)
W20190207-16:39:23.947(-8)? (STDERR)     at Server.call (packages/ddp-server/livedata_server.js:1603:17)
W20190207-16:39:23.947(-8)? (STDERR)     at ctor.<anonymous> (Routes/API.js:73:16)
W20190207-16:39:23.948(-8)? (STDERR)     at boundNext (packages/iron_middleware-stack.js:408:31)
W20190207-16:39:23.948(-8)? (STDERR)     at runWithEnvironment (packages/meteor.js:1356:24)
W20190207-16:39:23.948(-8)? (STDERR)     at packages/meteor.js:1369:14
W20190207-16:39:23.948(-8)? (STDERR)     at ctor.urlencodedParser (/Users/andrew/.meteor/packages/iron_router/.1.1.2.1kmiw8w.53mm++os+web.browser+web.cordova/npm/node_modules/body-parser/lib/types/urlencoded.js:77:44)
W20190207-16:39:23.949(-8)? (STDERR)     at packages/iron_router.js:886:36
W20190207-16:39:23.949(-8)? (STDERR)     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1304:12)
W20190207-16:39:23.949(-8)? (STDERR)     at ctor.hookWithOptions (packages/iron_router.js:885:27)
W20190207-16:39:23.950(-8)? (STDERR)     at boundNext (packages/iron_middleware-stack.js:408:31)
W20190207-16:39:23.950(-8)? (STDERR)     at runWithEnvironment (packages/meteor.js:1356:24)
```

**Comment by mommothazaz123** [comment]
Moving everything to Meteor methods and removing the changes to `parenting.js` and `characterAssetAllowDeny.js` still raises the same error when invoking the Meteor method. 
```
I20190207-22:05:07.957(-8)? Exception in callback of async function: { Error: No UserId set when trying to update character asset. [Access Denied, no userId]
I20190207-22:05:07.957(-8)?     at checkPermission (lib/functions/parenting.js:140:9)
I20190207-22:05:07.958(-8)?     at MethodInvocation.updateChildren (lib/functions/parenting.js:195:3)
I20190207-22:05:07.958(-8)?     at maybeAuditArgumentChecks (packages/ddp-server/livedata_server.js:1767:12)
I20190207-22:05:07.958(-8)?     at DDP._CurrentMethodInvocation.withValue (packages/ddp-server/livedata_server.js:1685:15)
I20190207-22:05:07.958(-8)?     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1304:12)
I20190207-22:05:07.958(-8)?     at resolve (packages/ddp-server/livedata_server.js:1683:36)
I20190207-22:05:07.958(-8)?     at new Promise (<anonymous>)
I20190207-22:05:07.958(-8)?     at Server.applyAsync (packages/ddp-server/livedata_server.js:1682:12)
I20190207-22:05:07.958(-8)?     at Server.apply (packages/ddp-server/livedata_server.js:1621:26)
I20190207-22:05:07.959(-8)?     at Server.call (packages/ddp-server/livedata_server.js:1603:17)
I20190207-22:05:07.959(-8)?     at Object.<anonymous> (lib/functions/parenting.js:115:10)
I20190207-22:05:07.959(-8)?     at packages/matb33_collection-hooks.js:486:20
I20190207-22:05:07.959(-8)?     at Array.forEach (<anonymous>)
I20190207-22:05:07.959(-8)?     at Function._.each._.forEach (packages/underscore.js:139:11)
I20190207-22:05:07.959(-8)?     at packages/matb33_collection-hooks.js:485:11
I20190207-22:05:07.959(-8)?     at Array.forEach (<anonymous>)
I20190207-22:05:07.960(-8)?  => awaited here:
I20190207-22:05:07.960(-8)?     at Promise.await (/Users/andrew/.meteor/packages/promise/.0.11.1.1calf8t.hslz++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/promise_server.js:60:12)
I20190207-22:05:07.960(-8)?     at Server.apply (packages/ddp-server/livedata_server.js:1634:14)
I20190207-22:05:07.960(-8)?     at Server.call (packages/ddp-server/livedata_server.js:1603:17)
I20190207-22:05:07.960(-8)?     at Object.<anonymous> (lib/functions/parenting.js:115:10)
I20190207-22:05:07.960(-8)?     at packages/matb33_collection-hooks.js:486:20
I20190207-22:05:07.960(-8)?     at Array.forEach (<anonymous>)
I20190207-22:05:07.961(-8)?     at Function._.each._.forEach (packages/underscore.js:139:11)
I20190207-22:05:07.961(-8)?     at packages/matb33_collection-hooks.js:485:11
I20190207-22:05:07.961(-8)?     at Array.forEach (<anonymous>)
I20190207-22:05:07.961(-8)?     at Function._.each._.forEach (packages/underscore.js:139:11)
I20190207-22:05:07.962(-8)?     at after (packages/matb33_collection-hooks.js:484:9)
I20190207-22:05:07.962(-8)?     at Object.<anonymous> (packages/matb33_collection-hooks.js:505:5)
I20190207-22:05:07.962(-8)?     at Object.collection.(anonymous function) [as update] (packages/matb33_collection-hooks.js:146:21)
I20190207-22:05:07.962(-8)?     at ns.Collection.update (packages/mongo/collection.js:585:31)
I20190207-22:05:07.962(-8)?     at ns.Collection.Mongo.Collection.(anonymous function) [as update] (packages/aldeed_collection2-core.js:232:19)
I20190207-22:05:07.962(-8)?     at inheritParentProperties (lib/functions/parenting.js:54:13)
I20190207-22:05:07.962(-8)?   isClientSafe: true,
I20190207-22:05:07.963(-8)?   error: 'Access Denied, no userId',
I20190207-22:05:07.963(-8)?   reason: 'No UserId set when trying to update character asset.',
I20190207-22:05:07.963(-8)?   details: undefined,
I20190207-22:05:07.963(-8)?   message: 'No UserId set when trying to update character asset. [Access Denied, no userId]',
I20190207-22:05:07.963(-8)?   errorType: 'Meteor.Error' }
```

**Comment by mommothazaz123** [comment]
Update 2: Looks like the initial inset works inside a Meteor method, but then a function gets called (`updateChildren`) which then calls `checkPermission`, which subsequently errors out

**Comment by ThaumRystra** [comment]
If you know your entire update is correct and no hooks need to be run to propagate changes around the character sheet, you can use `Collection.direct.insert` instead of `Collection.insert` to have your insert take place without any other hooks firing.

**Comment by mommothazaz123** [comment]
Okay, changed that. It should now validate that the objects the user gave it match the schema (adding default keys/values if required) before inserting. As far as I can tell, everything looks good; any other comments?

**Pull request merged**